File: Answer/67.二进制求和.go
/*
 * @lc app=leetcode.cn id=67 lang=golang
 *
 * [67] 二进制求和
 */
import (
	"fmt"
)
// @lc code=start
func addBinary(a string, b string) string {
    if len(b) > len(a) {
		a, b = b, a
	}
    lenA := len(a)
	lenB := len(b)
	diff := lenA - lenB
	
	result := make([]byte, lenA)
	carry := 0
	
	for i := lenB - 1; i >= 0; i-- {
		bitA := 0
		if a[i+diff] == '1' {
			bitA = 1
		}
		
		bitB := 0
		if b[i] == '1' {
			bitB = 1
		}
		
		sum := bitA + bitB + carry
		if sum > 1 {
			carry = 1
			sum -= 2
		} else {
			carry = 0
		}
		
		if sum == 1 {
			result[i+diff] = '1'
		} else {
			result[i+diff] = '0'
		}
	}
	
	for i := diff - 1; i >= 0; i-- {
		bitA := 0
		if a[i] == '1' {
			bitA = 1
		}
		
		sum := bitA + carry
		if sum > 1 {
			carry = 1
			sum -= 2
		} else {
			carry = 0
		}
		
		if sum == 1 {
			result[i] = '1'
		} else {
			result[i] = '0'
		}
	}
	
	resultStr := string(result)
	if carry == 1 {
		resultStr = "1" + resultStr
	}
	
	return resultStr
}
// @lc code=end

File: Answer/67.二进制求和.ts
/*
 * @lc app=leetcode.cn id=67 lang=typescript
 *
 * [67] 二进制求和
 */

// @lc code=start
function addBinary(a: string, b: string): string {
    function CtoI(c: string): number {
        if (c == '1') return 1;
        return 0;
    }

    function ItoC(i: number): string {
        if (i == 1) return '1';
        return '0';
    }

    let s = "";
    if (b.length > a.length) {
        s = a;
        a = b;
        b = s;
    }
    let c = [];
    let len = a.length - b.length;
    let t = 0;
    for (let i = b.length - 1; i >= 0; i--) {
        let aa = CtoI(a.charAt(i + len)) + CtoI(b.charAt(i)) + t;
        if (aa > 1) {
            t = 1;
            aa -= 2;
        } else {
            t = 0;
        }
        c[i + len] = ItoC(aa);
    }
    for (let i = len - 1; i >= 0; i--) {
        let aa = CtoI(a.charAt(i)) + t;
        if (aa > 1) {
            t = 1;
            aa -= 2;
        } else {
            t = 0;
        }
        c[i] = ItoC(aa);
    }
    s = c.toString().replace(/,/g, "");
    if (t == 1) {
        s = "1" + s;
    }
    return s;
};
// @lc code=end

